In [1]:
import numpy as np
from numpy import matrix
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LinearRegression

In [2]:
df1 = pd.read_csv("homes.csv")

In [3]:
df1.dtypes

price       int64
area        int64
beds        int64
baths       int64
garage      int64
year        int64
style       int64
lotsize     int64
ac         object
pool       object
quality    object
highway    object
dtype: object

In [563]:
df1["year"] = 2001 - df1["year"] # substitute year to age

In [564]:
# change the column name year to age
df1.rename(columns = {"year":"age"},
          inplace = True)

In [565]:
df1.head()

price  area  beds  baths  garage  age  style  lotsize   ac pool quality  \
0  360000  3032     4      4       2   29      1    22221  YES   NO  MEDIUM   
1  340000  2058     4      2       2   25      1    22912  YES   NO  MEDIUM   
2  250000  1780     4      3       2   21      1    21345  YES   NO  MEDIUM   
3  205500  1638     4      2       2   38      1    17342  YES   NO  MEDIUM   
4  275500  2196     4      3       2   33      7    21786  YES   NO  MEDIUM   

  highway  
0      NO  
1      NO  
2      NO  
3      NO  
4      NO

### Question1.  Fit a model with one predictor ( # of bedrooms), what is the estimated increase in average price?

In [566]:
beds = df1["beds"]
price = df1["price"]

In [567]:
beds.shape

(522,)

In [568]:
# reshape the beds

In [569]:
beds1 = beds.values.reshape(-1,1)
beds1.shape

(522, 1)

In [570]:
m1 = LinearRegression().fit(beds1,price) # first to write predictor(x)
m1.fit(beds1,price)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [571]:
# intercept and slope

In [572]:
m1.intercept_

82808.79961261933

In [573]:
m1.coef_

array([56200.08366568])

#### answer1: the estimated increase in the average of price is 56200 for each additional bedroom.

### Question 2 a)

In [574]:
import statsmodels.api as sm

In [575]:
X = df1[["lotsize","area","beds","baths","age","garage"]].to_numpy() # to show as a dataframe
y = df1[["price"]]
X = sm.add_constant(X) # to make the regression function to be eligible
m2 = sm.OLS(y,X).fit()
m2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.751
Model:                            OLS   Adj. R-squared:                  0.748
Method:                 Least Squares   F-statistic:                     259.4
Date:                Tue, 24 Sep 2019   Prob (F-statistic):          4.21e-152
Time:                        16:16:35   Log-Likelihood:                -6554.5
No. Observations:                 522   AIC:                         1.312e+04
Df Residuals:                     515   BIC:                         1.315e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -6707.5382   1.88e+04     -0.356      0.722   -4.37e+04    3.03e+04
x1             1.5550      0.272      5.717      0.000       1.021       2.089
x2           125.7386      6.963     18.057      0.000     112.059     139.419
x3         -1.304e+04   3780.397     -3.450      0.001   -2.05e+04   -5614.498
x4          7987.5518   4818.957      1.658      0.098   -1479.680    1.75e+04
x5         -1779.6108    215.487     -8.259      0.000   -2202.953   -1356.268
x6          2.253e+04   5814.344      3.875      0.000    1.11e+04     3.4e+04
==============================================================================
Omnibus:                      124.114   Durbin-Watson:                   1.327
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              384.582
Skew:                           1.110   Prob(JB):                     3.08e-84
Kurtosis:                       6.571   Cond. No.                     1.71e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.71e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [576]:
print("The coeffecients are:\n",m2.params) # to show all the coefficients
print("The estimated regression coefficients for a 99% confidence interval are:\n",m2.conf_int(alpha = 0.01))


The coeffecients are:
 const    -6707.538167
x1           1.554990
x2         125.738603
x3      -13041.394214
x4        7987.551827
x5       -1779.610838
x6       22530.379289
dtype: float64
The estimated regression coefficients for a 99% confidence interval are:
                   0             1
const -55431.280841  42016.204508
x1         0.851721      2.258260
x2       107.735683    143.741523
x3    -22815.268707  -3267.519722
x4     -4471.425313  20446.528968
x5     -2336.734049  -1222.487627
x6      7497.919937  37562.838640


### Question 2 b)

In [577]:
newval = pd.DataFrame([[1,24500,2650,3,3,11,2]],columns = ["const","lotsize","area","beds","baths","age","garage"])
newval

const  lotsize  area  beds  baths  age  garage
0      1    24500  2650     3      3   11       2

In [578]:
pred = m2.get_prediction(newval)
pred.summary_frame(alpha = 0.05)

mean      mean_se  mean_ci_lower  mean_ci_upper  obs_ci_lower  \
0  374920.532647  6511.365875  362128.426983   387712.63831  238431.17994   

    obs_ci_upper  
0  511409.885353

#### 2b) The average price for a 95% confidence interval is between 36128.43 and 387712.64.

### Question 2 c) price for house when all predictors are equal to ther median values.

In [579]:
list1 = ["lotsize","area","beds","baths","age","garage"]

In [580]:
medianP = df1[list1].median()

In [581]:
medianP

lotsize    22200.0
area        2061.0
beds           3.0
baths          3.0
age           35.0
garage         2.0
dtype: float64

In [582]:
lr = LinearRegression()
lr.fit(df1[list1],df1.price)
lr.predict(medianP.values.reshape(1,-1))

array([254573.35791857])

#### 2c) The price is 254573 when all predictors are equal to theri median.

### Question 3 best and worst predictors in terms of correlations

In [583]:
df1.corr()[:1]

price     area      beds     baths    garage       age     style  \
price    1.0  0.81947  0.413324  0.683685  0.577786 -0.555516  0.357493   

        lotsize  
price  0.224169

#### 3. Obvisouly, Area has 81.95% correlation with price and lotsize has 22.42% correlation with price.
####     So the best predictor is area and worst predictor is lotsize.

### Question 4 best and worst predictors interms of R-squared

In [584]:
from scipy.special import comb
from sklearn.metrics import mean_squared_error

In [585]:
Y1 = df1["price"]
X1 = df1[["lotsize","area","beds","baths","age","garage"]]

In [586]:
Y1.shape

(522,)

In [587]:
X1.shape

(522, 6)

In [588]:
model1 = LinearRegression().fit(X1,Y1)

In [589]:
MSE = mean_squared_error(Y1,model1.predict(X1))
MSE

4720225553.216769

In [590]:
SSE = MSE * len(Y1)
SSE

2463957738779.1533

In [591]:
R_squared = model1.score(X1,Y1)
R_squared

0.7513894013091613

In [592]:
# Another ways to get SSE and R-squared

In [593]:
from scipy.special import comb

In [594]:
def get_sse(X1,Y1):
    model_k = LinearRegression().fit(X1,Y1)
    SSE = mean_squared_error(Y1,model_k.predict(X1))*len(Y1)
    R_squared = model_k.score(X1,Y1)
    return SSE, R_squared

In [595]:
SSE, R_Squared = get_sse(X1,Y1)
SSE

2463957738779.1533

In [596]:
R_squared

0.7513894013091613

In [597]:
# create four empty list

In [598]:
SSE_list, R2_list, feature_list, num_features = [],[],[],[]

In [599]:
import itertools

In [600]:
m = len(X1.columns)
m

6

In [601]:
list(range(1,m+1))

[1, 2, 3, 4, 5, 6]

In [602]:
for k in range(1,m+1):
    for subset in itertools.combinations(X1.columns,k): # how many predictors
        temp = get_sse(X1[list(subset)],Y1)
    
        SSE_list.append(temp[0])
        R2_list.append(temp[1])
        feature_list.append(subset)
        num_features.append(len(subset))

In [603]:
df = pd.DataFrame(num_features,columns=["n_features"])

In [604]:
df["features"] = feature_list

In [605]:
df["R-squared"] = R2_list

In [606]:
SSE_list[:5]

[9412873469063.32,
 3255425605236.3877,
 8217765186285.018,
 5278296874758.204,
 6852419094443.968]

In [607]:
SSE_list2 = np.round(SSE_list,0)

In [608]:
df['SSE'] = SSE_list2

In [610]:
df[:13]

n_features           features  R-squared           SSE
0            1         (lotsize,)   0.050252  9.412873e+12
1            1            (area,)   0.671531  3.255426e+12
2            1            (beds,)   0.170837  8.217765e+12
3            1           (baths,)   0.467426  5.278297e+12
4            1             (age,)   0.308599  6.852419e+12
5            1          (garage,)   0.333837  6.602283e+12
6            2    (lotsize, area)   0.680802  3.163546e+12
7            2    (lotsize, beds)   0.200856  7.920250e+12
8            2   (lotsize, baths)   0.483056  5.123388e+12
9            2     (lotsize, age)   0.387781  6.067645e+12
10           2  (lotsize, garage)   0.352833  6.414020e+12
11           2       (area, beds)   0.674317  3.227813e+12
12           2      (area, baths)   0.681295  3.158659e+12

In [611]:
# group by n_feature

In [612]:
df.groupby(["n_features"])["R-squared"].max()

n_features
1    0.671531
2    0.718249
3    0.737944
4    0.745438
5    0.750063
6    0.751389
Name: R-squared, dtype: float64

In [613]:
df.groupby(["n_features"])["R-squared"].max()

n_features
1    0.671531
2    0.718249
3    0.737944
4    0.745438
5    0.750063
6    0.751389
Name: R-squared, dtype: float64

In [614]:
best_r2 = df.groupby(["n_features"])["R-squared"].max()

In [615]:
worst_r2 = df.groupby(["n_features"])["R-squared"].min()

In [616]:
# dataframe with best single predictor

In [620]:
df2 = df[df["R-squared"]== best_r2[1]]
df2

n_features features  R-squared           SSE
1           1  (area,)   0.671531  3.255426e+12

In [623]:
df2 = df[df['R-squared'] == best_r2[2]]
df2

n_features     features  R-squared           SSE
13           2  (area, age)   0.718249  2.792410e+12

In [624]:
# dataframes with best subsets of predictors

In [625]:
df2 = df[df["R-squared"] == best_r2[1]]

In [626]:
for i in range(2,7):
    df2 = df2.append(df[df['R-squared'] == best_r2[i]])

In [443]:
df2

n_features                                   features  R-squared  \
1            1                                    (area,)   0.671531   
13           2                                (area, age)   0.718249   
23           3                       (lotsize, area, age)   0.737944   
46           4               (lotsize, area, age, garage)   0.745438   
58           5         (lotsize, area, beds, age, garage)   0.750063   
62           6  (lotsize, area, beds, baths, age, garage)   0.751389   

             SSE  
1   3.255426e+12  
13  2.792410e+12  
23  2.597213e+12  
46  2.522941e+12  
58  2.477102e+12  
62  2.463958e+12

In [444]:
 #above is the forward selection to see the best subsets and AIC is the opposize, least is better

#### 4. From the data frame above, we can find that area is the best predictor, and baths is the worst predictor.

### 5. Fit model with best subset of predictors(AIC).

In [445]:
n = 522

In [447]:
df2['AIC'] = n* np.log(df2['SSE']/n) + 2* df2['n_features']

In [448]:
df2['adj_R-squared'] = 1 - ((1 - df2['R-squared'])*(n-1)/(n - df2['n_features'] - 1))

In [449]:
df2

n_features                                   features  R-squared  \
1            1                                    (area,)   0.671531   
13           2                                (area, age)   0.718249   
23           3                       (lotsize, area, age)   0.737944   
46           4               (lotsize, area, age, garage)   0.745438   
58           5         (lotsize, area, beds, age, garage)   0.750063   
62           6  (lotsize, area, beds, baths, age, garage)   0.751389   

             SSE           AIC  adj_R-squared  
1   3.255426e+12  11775.019159       0.670900  
13  2.792410e+12  11696.934938       0.717163  
23  2.597213e+12  11661.107762       0.736426  
46  2.522941e+12  11647.962610       0.743468  
58  2.477102e+12  11640.391242       0.747641  
62  2.463958e+12  11639.613904       0.748493

In [450]:
# AIC best model

In [452]:
df2[df2.AIC == df2.AIC.min()]

n_features                                   features  R-squared  \
62           6  (lotsize, area, beds, baths, age, garage)   0.751389   

             SSE           AIC  adj_R-squared  
62  2.463958e+12  11639.613904       0.748493

In [453]:
list2 = ["lotsize","area","beds","baths","age","garage"]
list2

['lotsize', 'area', 'beds', 'baths', 'age', 'garage']

In [454]:
Df_new = X1[list2]
Df_new[:5]

lotsize  area  beds  baths  age  garage
0    22221  3032     4      4   29       2
1    22912  2058     4      2   25       2
2    21345  1780     4      3   21       2
3    17342  1638     4      2   38       2
4    21786  2196     4      3   33       2

In [455]:
m_new = LinearRegression().fit(Df_new,Y1)

In [458]:
m_new.coef_

array([ 1.55499026e+00,  1.25738603e+02, -1.30413942e+04,  7.98755183e+03,
       -1.77961084e+03,  2.25303793e+04])

In [461]:
m_new.intercept_

-6707.538166642073

#### Answer 5

In [462]:
print("Price =",m_new.intercept_ ,"+",m_new.coef_[0],"*lotsize +",m_new.coef_[1],"*area +",m_new.coef_[2],"*beds +",m_new.coef_[3],"*baths+",m_new.coef_[4],"*age +",m_new.coef_[5],"*garage")

Price = -6707.538166642073 + 1.5549902587125097 *lotsize + 125.73860274607578 *area + -13041.394214126049 *beds + 7987.551827467522 *baths+ -1779.6108381452011 *age + 22530.379288592536 *garage


### Consider predictors lotsize and quality only

### 6.Fit a linear regression model and fitted equations

In [471]:
df3 = df1[["lotsize", "quality","price"]]
df3[:5]

lotsize quality   price
0    22221  MEDIUM  360000
1    22912  MEDIUM  340000
2    21345  MEDIUM  250000
3    17342  MEDIUM  205500
4    21786  MEDIUM  275500

In [472]:
df3["quality"].unique()

array(['MEDIUM', 'LOW', 'HIGH'], dtype=object)

In [473]:
# Encode categorical vars

In [474]:
price1 = pd.get_dummies(df3,columns = ["quality"],
                       drop_first = True)

In [476]:
price1[:15]

lotsize   price  quality_LOW  quality_MEDIUM
0     22221  360000            0               1
1     22912  340000            0               1
2     21345  250000            0               1
3     17342  205500            0               1
4     21786  275500            0               1
5     18902  248000            0               1
6     18639  229900            0               1
7     22112  150000            0               1
8     14321  195000            1               0
9     32358  160000            1               0
10    56639  190000            1               0
11    30595  559000            0               0
12    23172  535000            0               0
13    35351  525000            0               0
14    24025  299900            0               1

In [477]:
list3 = ["lotsize","quality_LOW","quality_MEDIUM"]

In [479]:
X_new = price1[list3]

In [480]:
Y_new = price1["price"]

In [481]:
M = LinearRegression().fit(X_new,Y_new)

In [483]:
M.coef_

array([ 1.64205257e+00, -3.61721638e+05, -2.66554186e+05])

In [484]:
M.intercept_

499607.81588552624

#### Answer 6

In [485]:
print("Price =",M.intercept_ ,"+",M.coef_[0],"*lotsize +",M.coef_[1],"*quality_LOW +",M.coef_[2],"*quality_MEDIUM")

Price = 499607.81588552624 + 1.6420525652315556 *lotsize + -361721.6383357747 *quality_LOW + -266554.1858577667 *quality_MEDIUM


### Question 7

In [488]:
newv = pd.DataFrame([[40000,0,1]],columns = ["lotsize","quality_LOW","quality_MEDIUM"])
newv

lotsize  quality_LOW  quality_MEDIUM
0    40000            0               1

In [495]:
price_new = M.intercept_ + M.coef_[0]*40000 + M.coef_[1]*0 + M.coef_[2]*1

In [496]:
price_new

298735.73263702175

#### Answer 7. We can predict that the price of a house is 298736.73 with lotsize 40000 square feet and median quality.

In [512]:
df3=df1[(df1.beds >= 2) & (df1.beds <= 4)] 
df3[:10]

price  area  beds  baths  garage  age  style  lotsize   ac pool quality  \
0  360000  3032     4      4       2   29      1    22221  YES   NO  MEDIUM   
1  340000  2058     4      2       2   25      1    22912  YES   NO  MEDIUM   
2  250000  1780     4      3       2   21      1    21345  YES   NO  MEDIUM   
3  205500  1638     4      2       2   38      1    17342  YES   NO  MEDIUM   
4  275500  2196     4      3       2   33      7    21786  YES   NO  MEDIUM   
5  248000  1966     4      3       5   29      1    18902  YES  YES  MEDIUM   
6  229900  2216     3      2       2   29      7    18639  YES   NO  MEDIUM   
7  150000  1597     2      1       1   46      1    22112  YES   NO  MEDIUM   
8  195000  1622     3      2       2   26      1    14321  YES   NO     LOW   
9  160000  1976     3      3       1   83      1    32358   NO   NO     LOW   

  highway  
0      NO  
1      NO  
2      NO  
3      NO  
4      NO  
5      NO  
6      NO  
7      NO  
8      NO  
9      NO

In [525]:
df3_new = pd.get_dummies(df3,columns = ["style","ac","pool","quality","highway"],
                        drop_first = True)

In [526]:
df3_new[:5]

price  area  beds  baths  garage  age  lotsize  style_2  style_3  style_4  \
0  360000  3032     4      4       2   29    22221        0        0        0   
1  340000  2058     4      2       2   25    22912        0        0        0   
2  250000  1780     4      3       2   21    21345        0        0        0   
3  205500  1638     4      2       2   38    17342        0        0        0   
4  275500  2196     4      3       2   33    21786        0        0        0   

   style_5  style_6  style_7  style_9  ac_YES  pool_YES  quality_LOW  \
0        0        0        0        0       1         0            0   
1        0        0        0        0       1         0            0   
2        0        0        0        0       1         0            0   
3        0        0        0        0       1         0            0   
4        0        0        1        0       1         0            0   

   quality_MEDIUM  highway_YES  
0               1            0  
1               1            0  
2               1            0  
3               1            0  
4               1            0

In [527]:
X3 = df3_new.drop(columns = "price", axis = 1)
Y3 = df3_new[["price"]]
X3 = sm.add_constant(X3) # to make the regression function to be eligible
m3 = sm.OLS(Y3,X3).fit()
m3.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.852
Model:                            OLS   Adj. R-squared:                  0.845
Method:                 Least Squares   F-statistic:                     135.9
Date:                Tue, 24 Sep 2019   Prob (F-statistic):          1.08e-163
Time:                        15:45:39   Log-Likelihood:                -5451.8
No. Observations:                 445   AIC:                         1.094e+04
Df Residuals:                     426   BIC:                         1.102e+04
Df Model:                          18                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
const           1.296e+05   2.67e+04      4.848      0.000    7.71e+04    1.82e+05
area             117.2176      7.739     15.146      0.000     102.006     132.429
beds           -2222.9820   4148.594     -0.536      0.592   -1.04e+04    5931.279
baths           9246.5631   4285.513      2.158      0.032     823.180    1.77e+04
garage          7423.7368   4914.563      1.511      0.132   -2236.074    1.71e+04
age            -1300.6190    199.782     -6.510      0.000   -1693.300    -907.938
lotsize            1.3283      0.238      5.584      0.000       0.861       1.796
style_2        -2.006e+04   8638.441     -2.322      0.021    -3.7e+04   -3080.425
style_3        -1.151e+04   8251.632     -1.395      0.164   -2.77e+04    4707.107
style_4          2.31e+04   1.72e+04      1.346      0.179   -1.06e+04    5.68e+04
style_5        -7407.6304   1.55e+04     -0.479      0.632   -3.78e+04     2.3e+04
style_6         -3.06e+04   1.51e+04     -2.029      0.043   -6.02e+04    -951.905
style_7        -4.664e+04   8514.836     -5.477      0.000   -6.34e+04   -2.99e+04
style_9        -9.094e+04   5.26e+04     -1.728      0.085   -1.94e+05    1.25e+04
ac_YES         -1075.0761   7501.411     -0.143      0.886   -1.58e+04    1.37e+04
pool_YES        2.163e+04   1.06e+04      2.042      0.042     813.263    4.25e+04
quality_LOW    -1.368e+05   1.45e+04     -9.433      0.000   -1.65e+05   -1.08e+05
quality_MEDIUM -1.363e+05   1.08e+04    -12.640      0.000   -1.58e+05   -1.15e+05
highway_YES     -3.18e+04   1.69e+04     -1.884      0.060    -6.5e+04    1380.295
==============================================================================
Omnibus:                       58.445   Durbin-Watson:                   1.344
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              269.978
Skew:                           0.449   Prob(JB):                     2.37e-59
Kurtosis:                       6.709   Cond. No.                     5.71e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.71e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [528]:
# fit the condition for the model

In [535]:
df4=pd.DataFrame({"constant":[1],"area":[3150],"beds":[2],"baths":[3],"garage":[2],"age":[5],"lotsize":[26250],"style_2":[0],"style_3":[0],
                  "style_4":[0],"style_5":[0],"style_6":[0],"style_7":[0],"style_9":[0],
  "ac_YES":[1],
  "pool_YES":[1],
  "quality_LOW":[0],
"quality_MEDIUM":[0],
  "highway_YES":[0]})
df4

constant  area  beds  baths  garage  age  lotsize  style_2  style_3  \
0         1  3150     2      3       2    5    26250        0        0   

   style_4  style_5  style_6  style_7  style_9  ac_YES  pool_YES  quality_LOW  \
0        0        0        0        0        0       1         1            0   

   quality_MEDIUM  highway_YES  
0               0            0

In [536]:
predictions = m3.get_prediction(df4)
predictions.summary_frame(alpha=0.05)

mean       mean_se  mean_ci_lower  mean_ci_upper   obs_ci_lower  \
0  585922.526796  15323.494616  555803.458271  616041.595321  479849.693761   

    obs_ci_upper  
0  691995.359832

#### 8 Answer. The prediction interval for the price of a house is between 479849 and 691995.(which is the prediction price for a specified condition)